### PR TITLE
Implement initial parser

### PR DESCRIPTION
**Pull request by doomspork**
Currently a work-in-progress.

### CODE REVIEW

File: lib/user_agent_parser.ex [hunk 1]
```diff
@@ -1,19 +1,25 @@
...
+      worker(Storage, []),
     ]
 
-    # See http://elixir-lang.org/docs/stable/elixir/Supervisor.html
-    # for other strategies and supported options
     opts = [strategy: :one_for_one, name: UserAgentParser.Supervisor]
     Supervisor.start_link(children, opts)
   end
+
+  def parse(user_agent) do
```
Comment (ybur-yug) [inline lib/user_agent_parser.ex]: For stuuff in this I prefer to only keep application configuration logic, but thats just a nitpick

File: lib/user_agent_parser/processor.ex [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    safe =
+      user_agent
+      |> String.trim
+
+    {:ok, safe}
+  end
+
+  def parse!(user_agent) do
+    {user_agents, os, devices} = Storage.list
+
```
Comment (ybur-yug) [inline lib/user_agent_parser/processor.ex]: pointless newline?

File: lib/user_agent_parser/processor.ex [hunk 3]
```diff
@@ -0,0 +1,29 @@
+defmodule UserAgentParser.Processor do
+
+  alias UserAgentParser.Storage
+
+  def sanitize(user_agent) do
+    safe =
+      user_agent
+      |> String.trim
+
```
Comment (ybur-yug) [inline lib/user_agent_parser/processor.ex]: Is there a preferred style for the newline here? I feel like its made clear that the return is happening due to the lack of indentation. Don't have a strong opinion, just curious.

File: lib/user_agent_parser/processor.ex (hunk 2 above)
Comment (doomspork) [inline lib/user_agent_parser/processor.ex]: I'm not done 😝

File: lib/user_agent_parser.ex (hunk 1 above)
Comment (doomspork) [inline lib/user_agent_parser.ex]: This is just a helper, the real work is happening elsewhere.  I personally think these make for a clean public API with `UserAgentPaser.parse("...")` vs something like `UserAgentParser.Parser.parse("...")`.

**Comment by doomspork** [comment]
@ybur-yug don't get too caught up in code that looks unfinished — it is 😀 .  I'd love to get your feedback on the structure of things as it's coming together.

**Comment by veverkap** [comment]
WHERE IS NAG?

**Comment by doomspork** [comment]
Damn good question @veverkap!  I need to add the config for .credo and get it wired up.  About to dig in now

**Comment by ybur-yug** [comment]
@doomspork my B, misunderstood when you initially told me to peek so I just was being thorough. It all looks rad! Ill keep my eyes on it

**Comment by doomspork** [comment]
No worries @ybur-yug, I wasn't clear 😀

File: test/fixtures/patterns.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  - regex: 'HbbTV/\d+\.\d+\.\d+ \( ;(LG)E ;NetCast 4.0'
+    os_v1_replacement: '2013'
+
+device_parsers:
+  - regex: '(wap[\-\ ]browser|maui|netfront|obigo|teleca|up\.browser|midp|Opera Mini)'
+    regex_flag: 'i'
+    device_replacement: 'Generic Feature Phone'
+    brand_replacement: 'Generic'
+    model_replacement: 'Feature Phone'
+
```
Comment (doomspork) [inline test/fixtures/patterns.yml]: Extra whitespace.

**Comment by doomspork** [comment]
Alrighty @veverkap & @ybur-yug, give 'er a go, should be done!  Feedback welcomed.

File: lib/user_agent_parser.ex [hunk 5]
```diff
@@ -1,19 +1,38 @@
 defmodule UserAgentParser do
+  @moduledoc """
```
Comment (doomspork) [inline lib/user_agent_parser.ex]: Sometimes coming up with these docs feels like a chore 😛

**Comment by doomspork** [comment]
So unfortunately I just discovered: https://github.com/romul/uap-elixir.  I'm not a huge fan of his implementation so I'm going to keep this but will require a repo/app name change, suggestions?

Spitballin':

+ ex_agent
+ ua_parser

**Comment by doomspork** [comment]
I have found others!  Still not sure I want to trash this work though...

+ https://github.com/tuvistavie/elixir-browser
+ https://github.com/elixytics/ua_inspector

**Comment by doomspork** [comment]
Ping @ybur-yug

**Comment by ybur-yug** [comment]
I like `ua_parser`. Simple and explanatory, and avoids the overused `ex` prefix

**Comment by doomspork** [comment]
Sounds good @ybur-yug.  I can make the changes to the project or if you want to get in some commits you can.

**Comment by ybur-yug** [comment]
I can do it later today

**Comment by doomspork** [comment]
I'll add you as a contributor @ybur-yug 👍

**Pull request merged**